go/sync2/batcher.go

```diff
@@ -0,0 +1,75 @@
+/*
+Copyright 2017 Google Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package sync2
+
+import (
+	"time"
+)
+
+// Batcher delays concurrent operations for a configurable interval in order to
+// batch them up or otherwise clock their operation to run concurrently.
+//
+// It is implemented as a channel of int32s. Each waiter blocks on the channel
+// from which it gets a sequentially increasing batch ID when the timer elapses.
+//
+// Hence a waiter is delayed for at most the batch interval.
+type Batcher struct {
+	interval time.Duration
+	queue    chan int32
+	waiters  AtomicInt32
+	nextID   AtomicInt32
+}
+
+// NewBatcher returns a new Batcher
+func NewBatcher(interval time.Duration) *Batcher {
+	return &Batcher{
+		interval: interval,
+		queue:    make(chan int32),
+		waiters:  NewAtomicInt32(0),
+		nextID:   NewAtomicInt32(0),
+	}
+}
+
+// Wait adds a new waiter to the queue and blocks until the next batch
+func (b *Batcher) Wait() int32 {
```

Comment:
I think `int32` is an implementation detail and having `int` in the API instead would be nicer, especially since you're adding it in a common package.

Go guarantees you that `int` fits at least `int32`, so just casting at the return should be fine.